pkg/textparse/lex.l

```diff
@@ -65,65 +55,46 @@ M     [a-zA-Z_:]
-<lstateValue>(NaN)                    l.val = math.Float64frombits(value.NormalNaN)
-                                      l.state = lstateTimestamp
-
-<lstateValue>[^\n \t\r]+              // We don't parse strictly correct floats as the conversion
-                                      // repeats the effort anyway.
-                                      l.val, l.err = strconv.ParseFloat(yoloString(l.b[l.vstart:l.i]), 64)
-                                      if l.err != nil {
-                                          return -1
-                                      }
-                                      l.state = lstateTimestamp
-
-<lstateTimestamp>[ \t]+               l.tstart = l.i
-<lstateTimestamp>{D}+                 ts, err := strconv.ParseInt(yoloString(l.b[l.tstart:l.i]), 10, 64)
-                                      if err != nil {
-                                          l.err = err
-                                          return -1
-                                      }
-                                      l.ts = &ts
-<lstateTimestamp>[\r\n]+              l.nextMstart = l.i
-                                      return 1
-<lstateTimestamp>\0                   return 1
+\0                                    return tEOF
+\n                                    l.state = sInit; return tLinebreak
+<*>[ \t]+                             return tWhitespace
+
+#[ \t]+                               l.state = sComment
+#                                     return l.consumeComment()
+<sComment>HELP[\t ]+                  l.state = sMeta1; return tHelp
+<sComment>TYPE[\t ]+                  l.state = sMeta1; return tType
+<sMeta1>{M}({M}|{D})*                 l.state = sMeta2; return tMName
+<sMeta2>{C}+                          l.state = sInit; return tText
+
+{M}({M}|{D})*                         l.state = sValue; return tMName
+<sValue>\{                            l.state = sLabels; return tBraceOpen
+<sLabels>{L}({L}|{D})*                return tLName
+<sLabels>\}                           l.state = sValue; return tBraceClose
+<sLabels>=                            l.state = sLValue; return tEqual
+<sLabels>,                            return tComma
+<sLValue>\"(\\.|[^\\"])*\"            l.state = sLabels; return tLValue
+<sValue>[^{ \t\n]+                    l.state = sTimestamp; return tValue
+<sTimestamp>{D}+                      return tTimestamp
+<sTimestamp>\n                        l.state = sInit; return tLinebreak
 
 %%
-    l.err = fmt.Errorf("no token found")
-    return -1
+    // Workaround to gobble up comments that started with a HELP or TYPE
+    // prefix. We just consume all characters until we reach a newline.
+    // This saves us from adding disproportionate complexity to the parser.
+    if l.state == sComment {
+        return l.consumeComment()
+    }
+    return tInvalid
+}
+
+func (l *lexer) consumeComment() token {
+    for c := l.cur(); ; c = l.next() {
+        switch c {
+        case 0:
```

Comment:
The zero byte is a valid UTF-8 character

Reply:
Fine, will add another hack to make that work.
Is this madness also carried into OpenMetrics' text format?

Reply:
Yes, nuls have always been valid UTF-8.

Reply:
To clarify the problem. EOF matching is not supported: https://github.com/cznic/lex/blob/master/lex.go#L74

The zero byte is used for that. Thus any matching on it is practically not possible unless we pick another `uint8` as the EOF signal.

Reply:
Byte 255 is not possible in UTF-8.

Reply:
Yes, but that doesn't help since we then couldn't distinguish between an invalid string and a valid EOF :) We've to pick one and work around it.

Reply:
Best to parse all valid input then. We'll gain an EOF marker with OpenMetrics that'll cover that case.